File: typescript/src/adapters/gpu-wrap.ts
/**
 * Serverless GPU handler wraps — Modal / RunPod / Replicate.
 *
 * Phase 2 GPU foundation Task 7. Mirrors python/src/dexcost/gpu_wrap.py +
 * compute-wrap.ts's shape.
 *
 * Each wrap is a thin decorator that:
 *   1. Creates a GpuAccountant and attaches it to the active task as `_gpu`.
 *   2. Times the handler with performance.now().
 *   3. Calls accountant.snapshotEndAndBuild(durationMs) at exit.
 *   4. Persists the dual events (1 gpu_cost with cost_pending=true + N
 *      gpu_utilization_signal) via the global tracker's buffer.
 *   5. Handler exceptions are re-raised AFTER events are persisted (the
 *      GPU-seconds were consumed and Modal/RunPod/Replicate bill failed
 *      invocations identically to successful ones — capture spec §6 case 7).
 *
 * When no dexcost task is in context the wrap is transparent (capture spec
 * §6 case 2 — anonymous compute never creates orphan events).
 *
 * Browser-safe: short-circuits the accountant work off-Node.
 */

import { randomUUID } from "node:crypto";
import { GpuAccountant, type GpuAccountantHooks } from "../core/gpu-accountant.js";
import { GpuRuntimeKind } from "../core/gpu-runtime.js";
import { getCurrentTask } from "../core/context.js";
import { getCloudEnv } from "../cloud-detect.js";
import { getTracker } from "../core/tracker.js";
import { createCostEvent } from "../core/models.js";
import type { Task } from "../core/models.js";

function _isNode(): boolean {
  return typeof process !== "undefined" && !!process.versions?.node;
}

// ─── Test hook: override the GpuAccountant hooks injected by the wrap ───────

let _hooksOverride: Partial<GpuAccountantHooks> | null = null;

/**
 * Test-only — pre-set the hooks every wrap-created GpuAccountant should
 * receive (used to stub NVML / cgroup primitives). Pass null to restore.
 */
export function _setGpuAccountantHooksForTests(
  hooks: Partial<GpuAccountantHooks> | null,
): void {
  _hooksOverride = hooks;
}

// ─── Persist helpers ────────────────────────────────────────────────────────

function _persistGpuEvents(
  task: Task,
  costDetails: Record<string, any> | null,
  signalEvents: Array<Record<string, any>> | null,
): void {
  let tracker;
  try {
    tracker = getTracker();
  } catch {
    return; // No global tracker → nothing to persist
  }
  try {
    if (costDetails !== null) {
      const ev = createCostEvent({
        eventId: randomUUID(),
        taskId: task.taskId,
        eventType: "gpu_cost",
        costUsd: 0, // back-filled at task finalize
        costConfidence: "unknown",
        isRetry: false,
        details: costDetails,
      });
      tracker.buffer.addEvent(ev);
    }
    if (signalEvents) {
      for (const sig of signalEvents) {
        const ev = createCostEvent({
          eventId: randomUUID(),
          taskId: task.taskId,
          eventType: "gpu_utilization_signal",
          costUsd: 0, // Decision #3 observability-only
          costConfidence: "unknown",
          isRetry: false,
          details: sig,
        });
        tracker.buffer.addEvent(ev);
      }
    }
  } catch {
    // Fail-silent per convention §9.
    // eslint-disable-next-line no-console
    console.warn("gpu_wrap failed to persist events");
  }
}

async function _timeAndCapture<R>(
  accountant: GpuAccountant,
  handler: () => Promise<R> | R,
): Promise<R> {
  accountant.snapshotStart();
  const t0 = performance.now();
  try {
    return await handler();
  } finally {
    const durationMs = Math.trunc(performance.now() - t0);
    try {
      const { costDetails, signalEvents } = accountant.snapshotEndAndBuild(durationMs);
      const task = getCurrentTask();
      if (task !== undefined) {
        _persistGpuEvents(task, costDetails, signalEvents);
      }
    } catch {
      // eslint-disable-next-line no-console
      console.warn("gpu_wrap event-build failed");
    }
  }
}

function _makeAccountant(runtime: GpuRuntimeKind): GpuAccountant {
  return new GpuAccountant(
    runtime,
    getCloudEnv(),
    _hooksOverride ?? undefined,
  );
}

// ─── Modal ──────────────────────────────────────────────────────────────────

export function wrapModalHandler<A extends any[], R>(
  handler: (...args: A) => Promise<R> | R,
): (...args: A) => Promise<R> {
  return async (...args: A): Promise<R> => {
    const task = getCurrentTask();
    if (!task) return await handler(...args);
    if (!_isNode()) return await handler(...args);
    const accountant = _makeAccountant(GpuRuntimeKind.Modal);
    (task as any)._gpu = accountant;
    return await _timeAndCapture(accountant, () => handler(...args));
  };
}

// ─── RunPod ─────────────────────────────────────────────────────────────────

export function wrapRunpodHandler<A extends any[], R>(
  handler: (...args: A) => Promise<R> | R,
): (...args: A) => Promise<R> {
  return async (...args: A): Promise<R> => {
    const task = getCurrentTask();
    if (!task) return await handler(...args);
    if (!_isNode()) return await handler(...args);
    const accountant = _makeAccountant(GpuRuntimeKind.RunPod);
    (task as any)._gpu = accountant;
    return await _timeAndCapture(accountant, () => handler(...args));
  };
}

// ─── Replicate ──────────────────────────────────────────────────────────────

export function wrapReplicateHandler<A extends any[], R>(
  handler: (...args: A) => Promise<R> | R,
): (...args: A) => Promise<R> {
  return async (...args: A): Promise<R> => {
    const task = getCurrentTask();
    if (!task) return await handler(...args);
    if (!_isNode()) return await handler(...args);
    const accountant = _makeAccountant(GpuRuntimeKind.Replicate);
    (task as any)._gpu = accountant;
    return await _timeAndCapture(accountant, () => handler(...args));
  };
}

File: go/adapters/gpu_wrap.go
// Serverless GPU handler wraps — Modal / RunPod / Replicate.
// Mirrors python/src/dexcost/gpu_wrap.py.
//
// Each wrap:
//  1. Looks up the active dexcost task in ctx; if absent, becomes a
//     transparent pass-through (capture spec §6 case 2).
//  2. Creates a core.GpuAccountant for the runtime + ambient CloudEnv.
//  3. Registers it in the global GpuAccountant registry so the tracker
//     finds it at task finalize.
//  4. Times the handler. On exit (including error), snapshots the end NVML
//     state and persists 1 gpu_cost event (cost_pending=true) + N
//     gpu_utilization_signal events. Handler errors are re-raised AFTER
//     events are persisted because the GPU-seconds were consumed regardless
//     (Modal et al. bill failed invocations the same as successes).
//
// Phase 1 generic-handler signature pattern carried forward verbatim.

package adapters

import (
	"context"
	"log"
	"time"

	"github.com/shopspring/decimal"
	"github.com/google/uuid"

	"github.com/DexwoxBusiness/dexcost-sdk/go/cloud"
	"github.com/DexwoxBusiness/dexcost-sdk/go/core"
)

// GPUEventBuffer is the minimal subset of core.Buffer that GPU wraps need.
type GPUEventBuffer interface {
	InsertEvent(e core.Event) error
}

var gpuEventBuffer GPUEventBuffer

// SetGPUEventBuffer wires the GPU wraps to an event sink. Called from
// dexcost.Init(); tests swap it directly.
func SetGPUEventBuffer(b GPUEventBuffer) { gpuEventBuffer = b }

// WrapModalGPUHandler instruments a Modal handler — emits 1 gpu_cost +
// N gpu_utilization_signal events around each invocation.
func WrapModalGPUHandler[T any, R any](fn func(context.Context, T) (R, error)) func(context.Context, T) (R, error) {
	return wrapGPU(core.GpuRuntimeModal, fn)
}

// WrapRunpodGPUHandler instruments a RunPod handler.
func WrapRunpodGPUHandler[T any, R any](fn func(context.Context, T) (R, error)) func(context.Context, T) (R, error) {
	return wrapGPU(core.GpuRuntimeRunpod, fn)
}

// WrapReplicateGPUHandler instruments a Replicate handler.
func WrapReplicateGPUHandler[T any, R any](fn func(context.Context, T) (R, error)) func(context.Context, T) (R, error) {
	return wrapGPU(core.GpuRuntimeReplicate, fn)
}

// wrapGPU is the generic per-runtime body shared by Modal / RunPod / Replicate.
func wrapGPU[T any, R any](
	runtime core.GpuRuntimeKind,
	fn func(context.Context, T) (R, error),
) func(context.Context, T) (R, error) {
	return func(ctx context.Context, evt T) (R, error) {
		task := core.GetCurrentTask(ctx)
		if task == nil {
			return fn(ctx, evt)
		}
		accountant := core.NewGpuAccountant(runtime, cloud.GetCloudEnv())
		core.RegisterGpuAccountant(task.TaskID.String(), accountant)
		return timeAndCaptureGPU(ctx, accountant, task.TaskID, fn, evt)
	}
}

// timeAndCaptureGPU runs fn while sampling NVML. On exit (including handler
// error) it persists the dual events. Handler errors propagate AFTER events
// are persisted — the GPU-seconds were consumed (capture spec §6 case 7).
func timeAndCaptureGPU[T any, R any](
	ctx context.Context,
	accountant *core.GpuAccountant,
	taskID uuid.UUID,
	fn func(context.Context, T) (R, error),
	evt T,
) (result R, retErr error) {
	accountant.SnapshotStart()
	t0 := time.Now()
	defer func() {
		duration := time.Since(t0).Milliseconds()
		// Fail-silent shell around event build + persist.
		func() {
			defer func() {
				if r := recover(); r != nil {
					log.Printf("[dexcost] gpu_wrap panic during event build: %v", r)
				}
			}()
			cost, signals := accountant.SnapshotEndAndBuild(duration)
			persistGPUEvents(taskID, cost, signals)
		}()
	}()
	return fn(ctx, evt)
}

// persistGPUEvents inserts the gpu_cost event (cost_pending=true) and the
// per-device gpu_utilization_signal events through the configured buffer.
// The pricing engine back-fills cost_usd at task finalize.
func persistGPUEvents(taskID uuid.UUID, cost map[string]any, signals []map[string]any) {
	buf := gpuEventBuffer
	if buf == nil {
		return
	}
	if cost != nil {
		ev := core.NewEvent(taskID, core.EventTypeGPUCost)
		ev.CostUSD = decimal.Zero
		ev.CostConfidence = core.CostConfidenceUnknown
		ev.Details = cost
		if err := buf.InsertEvent(ev); err != nil {
			log.Printf("[dexcost] gpu_wrap failed to persist gpu_cost: %v", err)
		}
	}
	for _, sig := range signals {
		ev := core.NewEvent(taskID, core.EventTypeGPUUtilizationSignal)
		ev.CostUSD = decimal.Zero
		ev.CostConfidence = core.CostConfidenceExact
		ev.Details = sig
		if err := buf.InsertEvent(ev); err != nil {
			log.Printf("[dexcost] gpu_wrap failed to persist gpu signal: %v", err)
		}
	}
}
